lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -222,42 +222,23 @@ OpFoldResult InvalidValueOp::fold(ArrayRef<Attribute> operands) {
 OpFoldResult AddPrimOp::fold(ArrayRef<Attribute> operands) {
   /// Any folding here requires a bitwidth extension.
 
-  // add(x, invalid) -> invalid
-  if (operands[1].dyn_cast_or_null<InvalidValueAttr>() ||
-      operands[0].dyn_cast_or_null<InvalidValueAttr>())
-    return InvalidValueAttr::get(getType());
```

Comment:
This will leave invalids unfolded. Should we keep the more aggressive behavior and fold "add(x, invalid)" to x, with the assumption that it is safe to treat the invalid as a zero?

Reply:
We should keep the aggressive behavior deriving from interpreting invalid as zero here. It would need to be a canonicalizer, though, due to needing to sign extend the non-invalid operand.  I.e., transform needs to be `add(x, invalid) -> pad(x, x.getWidth() + 1)`. I can see about adding the canonicalizers to the arithmetic ops. :+1:

Does that make sense to you?

Reply:
Oh right. Totally makes sense, also makes sense to split it out to another patch if you prefer!

Reply:
No need for another patch. 😄  I updated this PR with additional commits that does this.  The `add` and `sub` stuff is handled with DRRs for their canonicalizers.